libvips/foreign/cgifsave.c

```diff
@@ -251,90 +308,125 @@ vips_foreign_save_cgif_write_frame( VipsForeignSaveCgif *cgif )
 		p += 4;
 	}
 
-	/* Do we need to compute a new palette? Do it if the frame sum
-	 * changes.
-	 *
-	 * frame_checksum 0 means no current colourmap.
+	/* Do we need to compute a new palette? Do it if the palette changes.
 	 */
-	if( !cgif->global_colour_table ) {
-		gint64 checksum;
-
-		/* We need a checksum which detects colour changes, but
-		 * doesn't care about pixel ordering.
-		 *
-		 * Scale RGBA differently so that changes like [0, 255, 0] 
-		 * to [255, 0, 0] are detected.
-		 */
-		checksum = 0;
-		p = frame_bytes;
-		for( i = 0; i < n_pels; i++ ) {
-			checksum += p[0] * 1000;
-			checksum += p[1] * 100;
-			checksum += p[2] * 10;
-			checksum += p[3];
-
-			p += 4;
+	if( cgif->global_colour_table ) {
+		quantisation_result = cgif->quantisation_result;
+		lp = vips__quantise_get_palette( quantisation_result );
+	} else {
+		if( vips__quantise_image_quantize( cgif->input_image, cgif->attr, 
+			&quantisation_result ) ) { 
+			vips_error( class->nickname, 
+				"%s", _( "quantisation failed" ) );
+			return( -1 );
 		}
+		lp = vips__quantise_get_palette( quantisation_result );
 
-		if( cgif->frame_checksum == 0 ||
-			checksum != cgif->frame_checksum ) { 
-			cgif->frame_checksum = checksum;
+		static double pal_change_threshold = 64.0;
```

Comment:
This method seems to be similar to the way the `maxerror` parameter is handled right now: [cgifsave.c#L177](https://github.com/libvips/libvips/blob/6260a37136973f3d231d0e2c28af6190126c664e/libvips/foreign/cgifsave.c#L177)
We could make the threshold dynamic (e.g. use `maxerror` here) and set `maxerror` to a value > 0 by default.

Reply:
I don't think `maxerror` fits well here. For example, 64 is a good value for the inter-palette error while it is too large for the inter-frame error. Yet I agree that this number should be dynamic. Maybe we should rename `maxerror` to, for example, `iterframe_maxerror` and add a new param for inter-palette max error.